### ISSUE TITLE
PubSub: Subscription.Receive returns without an error

### ISSUE THREAD

**Issue by krzysztofdrys** [GoogleCloudPlatform/google-cloud-go#1059]
## Client
PubSub

```
[[projects]]
  name = "cloud.google.com/go"
  packages = [
    "compute/metadata",
    "iam",
    "internal/optional",
    "internal/version",
    "pubsub",
    "pubsub/apiv1",
    "pubsub/internal/distribution"
  ]
  revision = "777200caa7fb8936aed0f12b1fd79af64cc83ec9"
  version = "v0.24.0"
```
## Describe Your Environment
Alpine Docker, deployed in kubernetes in GKE

## Expected Behavior
According to documentation:
```
// Receive calls f with the outstanding messages from the subscription.
// It blocks until ctx is done, or the service returns a non-retryable error.
```

## Actual Behavior
`Receive` returns without an error after about 1h. 

## More info
This issue was introduced somewhere between 0.22 and 0.24. In other words, when I switch to 0.22, I cannot observe this problem.

```
[[projects]]
  name = "cloud.google.com/go"
  packages = [
    "compute/metadata",
    "iam",
    "internal/optional",
    "internal/version",
    "pubsub",
    "pubsub/apiv1",
    "pubsub/internal/distribution"
  ]
  revision = "056a55f54a6cc77b440b31a56a5e7c3982d32811"
  version = "v0.22.0"
```

Just to enusre you, that this is not a mistake on our part (e.g. cancelling the context). We are using pubsub client in our wrapper, which looks like this:
```
func (r *receiver) Start(ctx context.Context) error {
	err := r.subscription.Receive(ctx, r.receive)

	log := r.log.WithFields(logrus.Fields{
		"context_error":   ctx.Err(),
		"error":            err,
	})

	if err != nil {
		log.Infof("Receiver exited with error: '%v'", err)
	} else {
		log.Infof("Receiver exited")
	}

	return err
}
```
When `Receive` exists without an error we are observing logs like: 
```
018-07-11 15:44:26.000 CEST Receiver exited {
 insertId:  "5ac6hwg1xi03vt"  
 jsonPayload: {
  error:  null   
  message:  "Receiver exited"   
  context_error:  null   
  timestamp:  "2018-07-11T13:44:26.694911319Z"   
 }
 labels: {…}  
 logName:  "projects/ingrid-head/logs/service-popular-pickuppoint"  
 receiveTimestamp:  "2018-07-11T13:44:31.407873607Z"  
 resource: {…}  
 severity:  "INFO"  
 timestamp:  "2018-07-11T13:44:26Z"  
}
```

I can provide you with more details if you need them.

**Comment by jba** [comment]
I tried to reproduce, but after several hours my Receive is still running.

Looking at the diffs between the two versions, I don't see what has changed that could cause this.

Looking at the code itself, I can't find a code path that would return in a nil without a context cancellation or timeout somewhere.

Could you add some logging to the pubsub client so we can narrow this down? Suggestions:

- https://github.com/GoogleCloudPlatform/google-cloud-go/blob/v0.24.0/pubsub/subscription.go#L413
- https://github.com/GoogleCloudPlatform/google-cloud-go/blob/v0.24.0/pubsub/subscription.go#L485
- https://github.com/GoogleCloudPlatform/google-cloud-go/blob/v0.24.0/pubsub/subscription.go#L498

These are the only places in the code where a nil return is possible, as far as I can see. If you identify others, log those too and let me know.

**Comment by krzysztofdrys** [comment]
**TL;DR** This is this line: https://github.com/GoogleCloudPlatform/google-cloud-go/blob/v0.24.0/pubsub/subscription.go#L485

**Longer version:**

I created a branch with logs you requested: https://github.com/krzysztofdrys/google-cloud-go/tree/add_logging_to_v0.24.0

I used this branch in my service:
```
[[projects]]
  branch = "add_logging_to_v0.24.0"
  name = "cloud.google.com/go"
  packages = [
    "compute/metadata",
    "iam",
    "internal/optional",
    "internal/version",
    "pubsub",
    "pubsub/apiv1",
    "pubsub/internal/distribution"
  ]
  revision = "c3f6dfbc38e85841914399de7df44423bdc58df9"
  source = "https://github.com/krzysztofdrys/google-cloud-go.git"
```

After 1h (service started at `2018-07-13T12:44:51.362143942Z` and finished at `2018-07-13T13:45:02.085362886Z`) I got this log:
```
subscription.go#L486
```
which comes from here: https://github.com/krzysztofdrys/google-cloud-go/commit/c3f6dfbc38e85841914399de7df44423bdc58df9#diff-b4556a07a2b7818c407e847791b06c3bR486

**Comment by krzysztofdrys** [comment]
By the way: I will be away next week, but if you need additional debug info I think someone else from my team should be able to help you.

**Comment by jba** [comment]
I have a theory. See https://code-review.googlesource.com/c/gocloud/+/30430.

It would be great if you could patch that in and see if it works.

**Comment by jba** [comment]
Fixed by 275b871711d8709d68133bb0d128b3c1ea462bba.
